src/functor.jl

```diff
@@ -39,12 +39,39 @@ function _default_walk(f, x)
   re(map(f, func))
 end
 
+usecache(::AbstractDict, x) = isleaf(x) ? anymutable(x) : ismutable(x)
+usecache(::Nothing, x) = false
+
+# function _anymutable(x::T) where {T}
+#   ismutable(x) && return true
+#   fs = fieldnames(T)
+#   isempty(fs) && return false
+#   return any(f -> anymutable(getfield(x, f)), fs)
+# end
```

Comment:
Does this fail to constant fold sometimes?

Otherwise LGTM. About the weird cases, we could argue it's more conservative to not cache in both. A false positive seems much worse than a false negative here IMO. Asking uses of a higher-level `isleaf` to take on additional responsibility for caching is also fine. Incidentally, this is why I think extracting out caching from Functors and making callbacks handle memoization themselves would be nice.

Reply:
It fails to be instant on surprisingly simple functions, I didn't try to dig into why:
```
julia> @btime fun_anymutable((x=(1,2), y=3))
  min 36.500 ns, mean 38.715 ns (1 allocation, 32 bytes)
false

julia> @btime gen_anymutable((x=(1,2), y=3))
  min 0.001 ns, mean 0.014 ns (0 allocations)
false
```
Perhaps more surprisingly, the generated one is also not free e.g. here:
```
julia> @btime fun_anymutable($(Metalhead.ResNet()))
  min 275.685 ns, mean 323.217 ns (9 allocations, 320 bytes)
true

julia> @btime gen_anymutable($(Metalhead.ResNet()))
  min 147.536 ns, mean 161.010 ns (1 allocation, 32 bytes)
true
```
That contains `Chain([...])` which... should just stop the recursion?

Reply:
Smaller example, in which the number of layers seems to matter:

<details>

```
julia> model = Chain(
         Conv((3, 3), 1 => 16),                # 160 parameters
         Conv((3, 3), 16 => 16),               # 2_320 parameters
         Conv((3, 3), 16 => 32),               # 4_640 parameters
         Conv((3, 3), 32 => 32),               # 9_248 parameters
         Conv((3, 3), 32 => 64),               # 18_496 parameters
         Conv((3, 3), 64 => 64),               # 36_928 parameters
         Dense(16384 => 10),                   # 163_850 parameters
       );

julia> @btime fun_anymutable($model)
  min 327.851 ns, mean 448.404 ns (10 allocations, 3.17 KiB)
true

julia> @btime gen_anymutable($model)
  min 215.463 ns, mean 238.700 ns (8 allocations, 608 bytes)
true

julia> model = Chain(
         Conv((3, 3), 1 => 16),                # 160 parameters
         Conv((3, 3), 16 => 16),               # 2_320 parameters
         # Conv((3, 3), 16 => 32),               # 4_640 parameters
         # Conv((3, 3), 32 => 32),               # 9_248 parameters
         # Conv((3, 3), 32 => 64),               # 18_496 parameters
         Conv((3, 3), 64 => 64),               # 36_928 parameters
         Dense(16384 => 10),                   # 163_850 parameters
       );

julia> @btime fun_anymutable($model)
  min 344.818 ns, mean 391.967 ns (10 allocations, 1.75 KiB)
true

julia> @btime gen_anymutable($model)
  min 0.001 ns, mean 0.014 ns (0 allocations)
true
```

</details>

Reply:
For the Metalhead example at least, that one allocation is coming from https://github.com/FluxML/Metalhead.jl/blob/7827ca6ec4ef7c5e07d04cd6d84a1a3b11289dc0/src/convnets/resnets/resnet.jl#L17. For the longer Chain, Cthulhu tells me
```
┌ Info: Inference didn't cache this call information because of imprecise analysis due to recursion:
└ Cthulhu nevertheless is trying to descend into it for further inspection.
```
If I add a guard against the possible `missing` from `any` in `gen_anymutable` and assert the return value like so:
```julia
@generated function gen_anymutable(x::T) where {T}
  ismutabletype(T) && return true
  fs = fieldnames(T)
  isempty(fs) && return false
  subs =  [:(gen_anymutable(getfield(x, $f))) for f in QuoteNode.(fs)]
  return :(coalesce(|($(subs...)), false)::Bool)
end
```
That eliminates all but 6 of the allocations. I believe these correspond to the 6 Conv layers because the check on the Dense layer appears to be fully const folded (why only the Dense? Not sure).

Reply:
Oh nice. Just the `::Bool` seems to be enough, and should be safe I think. 

Weirdly it is instant for 5 and 7 conv layers, only exactly 6 causes it to fail & take 100ns.

Reply:
That is absolutely bizarre. It also works for 6 Conv layers if I remove the final Dense and up to at least 32 with/without. Granted, this is on nightly&mdash;I couldn't get close to your timings on 1.8.2 IIRC.

Reply:
Ok, less bizarre. Trying to simplify things a bit, it looks like the _first_ call is always taking a hit here, but subsequent calls are fine.
```julia
using BenchmarkTools

struct Conv{N,M,F,A,V}
  σ::F
  weight::A
  bias::V
  stride::NTuple{N,Int}
  pad::NTuple{M,Int}
  dilation::NTuple{N,Int}
  groups::Int
end

struct Dense{F, M, B}
  weight::M
  bias::B
  σ::F
end

@generated function anymutable(x::T) where {T}
  ismutabletype(T) && return true
  fs = fieldnames(T)
  isempty(fs) && return false
  subs =  [:(anymutable(getfield(x, $f))) for f in QuoteNode.(fs)]
  return :(|($(subs...))::Bool)
end

function test()
  for N in (5, 6, 7)
    @info N
    layers = ntuple(_ -> Conv(identity, ones(1), ones(1), (1,), (1,), (1,), 1), N)
    layers = (layers..., Dense(ones(1), ones(1), identity))
    @btime anymutable($layers)
  end
end

test()
```
Perhaps that has something to do with the generated function?

Reply:
It's weird. 

I don't suggest doing this, but this does seem to compile away:
```
Base.@assume_effects :total function anymutable_total(x::T) where {T}
  ismutable(x) && return true
  fs = fieldnames(T)
  isempty(fs) && return false
  return any(f -> anymutable_total(getfield(x, f)), fs)::Bool
end
```

Reply:
Hmm, that still allocates (more, as a matter of fact) for me on nightly and 1.8. At least performance is consistent though.
```
[ Info: 5
  176.931 ns (7 allocations: 1.05 KiB)
[ Info: 6
  177.480 ns (7 allocations: 1.23 KiB)
[ Info: 7
  188.679 ns (7 allocations: 1.34 KiB)
```